bootstrap/kubeadm/controllers/kubeadmconfig_controller.go

```diff
@@ -469,7 +502,15 @@ func (r *KubeadmConfigReconciler) joinWorker(ctx context.Context, scope *Scope)
 	return ctrl.Result{}, nil
 }
 
-func (r *KubeadmConfigReconciler) joinControlplane(ctx context.Context, scope *Scope) (ctrl.Result, error) {
+func (r *KubeadmConfigReconciler) joinControlplane(ctx context.Context, scope *Scope) (_ ctrl.Result, reterr error) {
+	defer func() {
+		if reterr != nil {
+			// NOTE: We are surfacing join error because in some cases those are originated by misconfiguration
+			// that can be fixed by the user. e.g. invalid file source.
+			conditions.MarkFalse(scope.Config, bootstrapv1.DataSecretGeneratedCondition, bootstrapv1.DataSecretGeneratedFailedReason, clusterv1.ConditionSeverityWarning, reterr.Error())
+		}
+	}()
+
```

Comment:
Would it make sense to push this up to the Reconcile() method rather than handling through defer?

Reply:
same as above (marking as resolved so we track in a single point)